src/context.rs

```diff
@@ -318,9 +318,9 @@ unsafe impl<'buf> Context for AllPreallocated<'buf> {
     }
 }
 
-impl<'buf, C: Context + 'buf> Secp256k1<C> {
+impl<'buf> PreallocatedContext<'buf> for Secp256k1<AllPreallocated<'buf>> {
```

Comment:
Two comments; we can macroize this and also the trait needs to be public.
```rust
macro_rules! impl_preallocated_gen_new {
	($ty:ty) => {
		/// Lets you create a context with a preallocated buffer in a generic manner (sign/verify/all).
		fn preallocated_gen_new(buf: &'buf mut [AlignedType]) -> Result<Self, Error> {
			#[cfg(target_arch = "wasm32")]
			ffi::types::sanity_checks_for_wasm();

			if buf.len() < Self::preallocate_size_gen() {
				return Err(Error::NotEnoughMemory);
			}
			Ok(Secp256k1 {
				ctx: unsafe {
					NonNull::new_unchecked(ffi::secp256k1_context_preallocated_create(
						buf.as_mut_c_ptr() as *mut c_void,
						<$ty>::FLAGS,
					))
				},
				phantom: PhantomData,
			})
		}
	}
}
impl<'buf> PreallocatedContext<'buf> for Secp256k1<AllPreallocated<'buf>> {
	impl_preallocated_gen_new!(AllPreallocated<'buf>);
}
impl<'buf> PreallocatedContext<'buf> for Secp256k1<VerifyOnlyPreallocated<'buf>> {
	impl_preallocated_gen_new!(VerifyOnlyPreallocated<'buf>);
}
impl<'buf> PreallocatedContext<'buf> for Secp256k1<SignOnlyPreallocated<'buf>> {
	impl_preallocated_gen_new!(SignOnlyPreallocated<'buf>);
}
```

And, an a lighter note, this PR shows who round here is the C ludite that puts definitions above where they are used. I say that with much endearment and kindness as one often called a ludite to another :)

Reply:
100% agree. You could also macroize the `impl...` line :)

Also notably we **could** keep inherent method just forwarding to trait. This would keep the currently-sound client code working after upgrade and avoid annoying imports.

Reply:
I also realized none of this duplication/macro stuff would be needed if `PreallocatedContext` was an `unsafe` marker trait. Any specific reason you didn't choose it @apoelstra ?

Reply:
I figured, making stuff `unsafe` that didn't use to be would be a more annoying change for users than doing something ugly. I don't feel strongly about this.

We could actually leave the code alone except to mark the existing `preallocated_gen_new` funciton unsafe. And/or we could just make it private, and insist users go through the more specific `Secp256k1::<AllPreallocated<'a>>::preallocated_new` etc methods.

Reply:
`unsafe` trait is `unsafe to implement but safe to use. The API would be almost identical to the current, it'd just disallow unsound usage. So if someone used it soundly, without any generic crazyness it would work without change. Guaranteed UB would be guaranteed to fail to compile. Funny generics would have to be adjusted slightly but work as before.

Reply:
Ah! I'm sorry, I misunderstood what you were suggesting.

Now I understand you to be suggesting: (1) add a new `unsafe` marker trait `PreallocatedContext<'a>`; (2) impl this for the three preallocated context markes e.g. `AllPreallocated<'a>`; (3) change the bound on `preallocated_gen_new` from `C: Context + 'buf` to `C: Context + PreallocatedContext<'buf>`.

I believe this would work. I'll try it.

Reply:
Exactly.